### PR TITLE
CLOUDP-227273: Translation layer on Project IP Access List + state machine

### PR DESCRIPTION
**Pull request by helderjs**
### All Submissions:

* [x] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?
* [ ] Put `closes #XXXX` in your comment to auto-close the issue that your PR fixes (if there is one).
* [ ] Update docs/release-notes/release-notes-template.md if your changes should be included in the release notes for the next release.

### CODE REVIEW

File: internal/translation/ipaccesslist/conversion.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+
+func (i *IPAccessEntry) ID() string {
+	if i.CIDR != "" {
+		return i.CIDR
+	}
+
+	return i.AWSSecurityGroup
+}
+
+func (i *IPAccessEntry) IsExpired() bool {
```
Comment (s-urbaniak) [inline internal/translation/ipaccesslist/conversion.go:30]: for better testability:
```suggestion
func (i *IPAccessEntry) IsExpired(now time.Time) bool {
```

File: internal/translation/ipaccesslist/conversion.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+			Mask: net.IPv4Mask(255, 255, 255, 255),
+		}
+		return ipNet.String(), nil
+	}
+
+	if cidr != "" {
+		var err error
+		_, parsedNet, err := net.ParseCIDR(cidr)
+		if err != nil {
+			return "", fmt.Errorf("cidr %s is invalid", cidr)
```
Comment (s-urbaniak) [inline internal/translation/ipaccesslist/conversion.go:185]: return the error as well

**Review by s-urbaniak — APPROVED** [review]
this is fantastic 👏

File: internal/translation/ipaccesslist/conversion.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+	for ix, entry := range i {
+		if !entry.IsExpired() {
+			entries[ix] = entry
+		}
+	}
+
+	return entries
+}
+
+func (i IPAccessEntries) GetExpired() IPAccessEntries {
```
Comment (s-urbaniak) [inline internal/translation/ipaccesslist/conversion.go:51]:
```suggestion
func (i IPAccessEntries) ByExpired(expired bool) IPAccessEntries {
```

File: internal/translation/ipaccesslist/conversion.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	if i.DeleteAfterDate == nil {
+		return false
+	}
+
+	return i.DeleteAfterDate.Before(time.Now())
+}
+
+type IPAccessEntries map[string]*IPAccessEntry
+
+func (i IPAccessEntries) GetActives() IPAccessEntries {
```
Comment (s-urbaniak) [inline internal/translation/ipaccesslist/conversion.go:40]: this method can be removed in favor of a single `ByExpired(expired bool)` method (see below)

File: internal/translation/ipaccesslist/conversion.go [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+	for ix, entry := range i {
+		if entry.IsExpired() {
+			entries[ix] = entry
+		}
+	}
+
+	return entries
+}
+
+func ToAKO(ipAccessEntries IPAccessEntries) []project.IPAccessList {
```
Comment (s-urbaniak) [inline internal/translation/ipaccesslist/conversion.go:62]: I believe the convention elsewhere was `fromAtlas` and `toAtlas`. Let's stick to a common scheme.

**Review by s-urbaniak — CHANGES_REQUESTED** [review]
a couple of nits, else lgtm

File: internal/translation/ipaccesslist/conversion.go (hunk 5 above)
Comment (s-urbaniak) [reply internal/translation/ipaccesslist/conversion.go:62]: `func FromInternal(ipAccessEntries IPAccessEntries) []project.IPAccessList {`

Comment (s-urbaniak) [reply internal/translation/ipaccesslist/conversion.go:62]: @helderjs do you mind to add the "source signifier" convention into the translation layer proposal, including the rationale (aka the "why").

File: internal/translation/ipaccesslist/conversion.go (hunk 1 above)
Comment (josvazg) [reply internal/translation/ipaccesslist/conversion.go:30]: Nit pick I would call is `isExpired(at time.Time)` were usually it **at** is **now**: `.IsExpired(time.now())`

File: internal/translation/ipaccesslist/conversion.go (hunk 3 above)
Comment (josvazg) [reply internal/translation/ipaccesslist/conversion.go:51]: I see the value on the boolean arg, but the `GetByExpired` name sounds weird to me. Not sure what that means. `GetByStatus(expired bool)` seems clearer.

File: internal/translation/ipaccesslist/conversion.go (hunk 5 above)
Comment (josvazg) [reply internal/translation/ipaccesslist/conversion.go:62]: Yes, I thought we would only use 'toAtlas` and `fromAtlas` everywhere.

We also intend to overuse the AKO acronym in our code a lot.

File: pkg/controller/atlasproject/ipaccess_list.go [hunk 6]
```diff
@@ -1,207 +1,133 @@
...
+			}
+		}
 	}
 
-	desiredMap := map[string]project.IPAccessList{}
-	for _, item := range desired {
-		desiredMap[genIPAccessListKey(item)] = item
+	if isUnset {
+		return i.unmanage()
 	}
```
Comment (josvazg) [inline pkg/controller/atlasproject/ipaccess_list.go:65]: I find it confusing this `isUnset -> unmanage` is both here and in the main `reconcile` body. Can you elaborate on why is that needed?

**Review by josvazg — COMMENTED** [review]
LGTM overall. Just a question and no additional concerns apart from those by @s-urbaniak

File: pkg/controller/atlasproject/ipaccess_list.go (hunk 6 above)
Comment (helderjs) [reply pkg/controller/atlasproject/ipaccess_list.go:65]: This ensures status conditions are removed when config is removed from the manifest

File: internal/translation/ipaccesslist/conversion.go [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+	for ix, entry := range i {
+		if entry.IsExpired(time.Now()) == expired {
+			entries[ix] = entry
+		}
+	}
+
+	return entries
+}
+
+func FromInternal(ipAccessEntries IPAccessEntries) []project.IPAccessList {
```
Comment (josvazg) [inline internal/translation/ipaccesslist/conversion.go:52]: I was not understanding why we needed this, but then I saw we are changing the shape of the IP access list from an array to a set. That is fine, I guess it simplifies something later like checking for no duplicates.

Still most other "internal" types augment information that was not there in the spec, such as the project id or a password value.

To avoid the need for this `FromInternal` you could use the composition trick again:

```go
type IPAccessEntries struct {
  List []project.IPAccessList
  Set map[string]*IPAccessEntry
}
```

When passing the type you just need to use the `.List` instead of calling `FromInternal`. It is the original list, so no bugs possible.

Comment (josvazg) [reply internal/translation/ipaccesslist/conversion.go:52]: I saw later that this is used after a filtering call, so the trick will not work.

**Review by josvazg — APPROVED** [review]

**Review by s-urbaniak — APPROVED** [review]
